### ISSUE TITLE
Passing providers explicitly to terraform modules

### ISSUE THREAD

**Issue by muresan** [gruntwork-io/terragrunt#1290]
Going over documentation I could not find any options to use this feature from terraform via terragrunt. I've tried to add provider in inputs but to no avail.

https://www.terraform.io/docs/configuration/modules.html#passing-providers-explicitly

I'm talking here specifically about pasing the provider block to a module:

```
module "example" {
  source    = "./example"
  providers = {
    aws = "aws.usw2"
  }
}
```

I've tried something like:

```
terraform {
  source = "./example"
}

inputs = {
  providers = {
    google = "google-beta"
  }
}
```

the result was:

```
[terragrunt] 2020/08/06 13:50:16 Running command: terraform providers
.
└── provider.google
```

I expected to see provider.google-beta. There is no provider block in the module itself, it has a simple 'data' resource.

thank you for any pointers.

**Comment by yorinasub17** [comment]
The canonical pattern for this is to generate the provider config using [generate blocks](https://terragrunt.gruntwork.io/docs/features/keep-your-terraform-code-dry/#dry-common-terraform-code-with-terragrunt-generate-blocks). Note that this only works if the module doesn't have a provider config hard coded in there.

**Comment by muresan** [comment]
Just to follow up on my question: currently terraform does not offer a way to pass provider blocks to modules via variables or other parameters, so terragrunt cannot help. The only option here would be if we generate a block that includes the module call, but that defeats the purpose.

To reply to the comment, generate blocks will instantiate the provider, so you would see google-beta as the provider, but you cannot make the module itself use the provider, each resource that uses it has to explicitly require the module, so in this case adding a providers block is not really needed as terraform will add the google-beta provider if a resource requires it.

To sum up, there's currently no way to pass on the 
```
  providers = {
    aws = "aws.usw2"
  }
``` 
parameter to a module as it's not really a parameter and `providers` is not really a terraform variable. Maybe terraform needs something similar to -backend-config but for modules+providers combo.

**Comment by tomasbackman** [comment]
@muresan I had the same problem and asked the same question.. but got a working workaround as an answer. Basically it is possible to use an intermediate main.tf file and from there call the actual module source.
See: https://github.com/gruntwork-io/terragrunt/issues/1377

**Comment by dylan-shipwell** [comment]
modules that need two aliases of a provider cannot be used from a top level terragrunt file? did i read that right?
